File: 525-contiguous-array/leochang/525.cpp
class Solution {
public:
	int findMaxLength(vector<int>& nums) {
		unordered_map<int,int> map;
		int length = 0;
		int sum = 0;
		map[0] = -1;

		for (int i = 0; i < nums.size(); i++) {
			sum += nums[i] ? 1 : -1;
			if (map.end() == map.find(sum)) {
				map[sum] = i;
			} else {
				length = max(length, i - map[sum]);
			}
		}

		return length;
	}
};

File: 525-contiguous-array/dennis/525.py
class Solution(object):
    def findMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        max_length = 0
        count = 0
        
        count_idx_dict = {0: -1}
        
        for i in xrange(len(nums)):
            if nums[i] == 1:
                count += 1
            else:
                count -= 1
                
            if count in count_idx_dict:
                max_length = max(max_length, i - count_idx_dict[count])
            else:
                count_idx_dict[count] = i
                
        return max_length
